### PR TITLE
[Playground] support for Kafka-enabled examples

### PR DESCRIPTION
**Pull request by eantyshev**
This PR enables Kafka consumer logic in Beam examples. Playground server pre-loads (via Kafka producer) dataset data in scope of RunCode call.
Datasets content is stored under playground/backend/datasets/ to be easily accessed for development, but in future we could add support for larger datasets backed by public GCS buckets
It also requires additional runtime parsing of `bootstrap.server` in RunCode, to replace `kafka_server:9292` with actual host/port, `localhost:12345` or like. This logic is only implemented for Java. Kafka support in Python/Go SDK requires cross-language transforms implemented first [link](https://github.com/akvelon/beam/blob/b3963b75d3555a143a2e95ae913e956624ad5cc6/sdks/python/apache_beam/io/kafka.py).

### Examples layout
`beam-playground` embedded yaml tag is extended with `datasets` and `emulators`, for these examples appropriate pg_datasets entities are created
See more detailed example in CONTRIBUTE.md

### Playground GRPC API extensions
`datasets` in `RunCode` request and in `PrecompiledObject` message, returned by `GetPrecompiledObjects`
Client flow:
- receive example's datasets in a PrecompiledObject. Represent it nicely
- When RunCode is called, fill in `datasets` with a given struct

### Kafka Mock Cluster
under-hood implementation based on confluent-kafka-go client and its [librdkafka CGo bindings](https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_mock.h) to create MockCluster which dynamically opens port (2 bootstrapServers by default), dynamically creates a topic with 4 (by default) partitions when consumer is connected. (Beam example execution connects to that dynamically opened port).
Only consumer logic is implemented for now. To allow producing into a topic from Beam example, we need to extend API with returned artifacts to represent them safely for user.

### Tasks
addresses #23980
addresses #23981
addresses #23799

### Follow ups
- GCS-backed datasets
- Cross-language transforms
- Python SDK support with example
- Go SDK support with example
- Producing from Kafka Beam examples

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/containers/router/Dockerfile [hunk 1]
```diff
@@ -24,35 +24,22 @@ FROM $BASE_IMAGE AS build
 ENV GOPATH /go
 ENV PATH $GOPATH/bin:$PATH
 RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
-RUN go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28.1 &&\
-    go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2.0
 
 # Prepare Application
 COPY src /go/src/playground/backend
```
Comment (eantyshev) [inline playground/backend/containers/router/Dockerfile:30]: below are some remainings of Go SDK runner, not needed for the router component

File: playground/backend/containers/router/Dockerfile [hunk 2]
```diff
@@ -24,35 +24,22 @@ FROM $BASE_IMAGE AS build
...
-RUN go mod download &&\
-    go mod tidy &&\
+RUN go mod tidy -v &&\
+    go mod download &&\
     cd cmd/server &&\
     go build -o /go/bin/server_go_backend
 
 # Null image
-FROM debian:stretch-slim
+FROM debian:stable-20221114-slim
```
Comment (eantyshev) [inline playground/backend/containers/router/Dockerfile:42]: we need to keep it updated when Kafka client version is updated

File: playground/infrastructure/checker.py [hunk 3]
```diff
@@ -99,7 +99,7 @@ def main():
     if check_sdk_examples(args.paths, Sdk.Value(args.sdk), root_dir):
         return
```
Comment (eantyshev) [inline playground/infrastructure/checker.py:101]: exit code 1 can mask away checker.py exceptions, need to choose a somewhat more rare one

File: .github/workflows/playground_examples_ci.yml [hunk 4]
```diff
@@ -34,7 +34,7 @@ jobs:
   check:
     strategy:
       matrix:
-        sdk: ["python", "java", "go"]
+        sdk: ["java", "python", "go"]
```
Comment (eantyshev) [inline .github/workflows/playground_examples_ci.yml:37]: Java SDK is the most critical, to get CI results sooner

File: playground/infrastructure/datastore_client.py [hunk 5]
```diff
@@ -83,54 +85,65 @@ def save_to_cloud_datastore(self, examples_from_rep: List[Example], sdk: Sdk, or
         examples_ids_before_updating = self._get_all_examples(sdk, origin)
 
         # loop through every example to save them to the Cloud Datastore
```
Comment (eantyshev) [inline playground/infrastructure/datastore_client.py:87]: this huge chunk is only to remove transaction when updating examples: datastore transactions are limited to 500 entities affected, here we have possibly more.
https://cloud.google.com/datastore/docs/concepts/limits

**Review by damccorm — COMMENTED** [review]
Mostly looks good, just had a few comments

File: playground/infrastructure/datastore_client.py (hunk 5 above)
Comment (eantyshev) [reply playground/infrastructure/datastore_client.py:86]: The only reasonable restriction is to not leave single example-related entities in inconsistent state
I've refactored this place to reduce transaction scope to a single example

File: playground/backend/internal/db/mapper/precompiled_object_mapper_test.go [hunk 6]
```diff
@@ -41,7 +41,11 @@ func TestPrecompiledObjectMapper_ToObjectInfo(t *testing.T) {
 		actualResult.ContextLine != 32 ||
 		len(actualResult.Categories) != 3 ||
 		actualResult.Type.String() != "PRECOMPILED_OBJECT_TYPE_EXAMPLE" ||
-		actualResult.Sdk != pb.Sdk_SDK_JAVA {
+		actualResult.Sdk != pb.Sdk_SDK_JAVA ||
+		len(actualResult.Datasets) != 1 ||
+		actualResult.Datasets[0].DatasetPath != "MOCK_PATH_0" ||
+		actualResult.Datasets[0].Options["Topic"] != "MOCK_TOPIC" ||
+		actualResult.Datasets[0].Type != pb.EmulatorType_EMULATOR_TYPE_KAFKA {
 		t.Error("ToObjectInfo() unexpected result")
```
[the remark this answers is not in the document]
Comment (eantyshev) [reply playground/backend/internal/db/mapper/precompiled_object_mapper_test.go]: ++ I changed this for assert.Equal to show detailed diff
there're many more such places in this file though, couldn't fix it all

[the remark this answers is not in the document]
Comment (damccorm) [reply playground/backend/internal/db/mapper/precompiled_object_mapper_test.go]: SGTM

File: playground/infrastructure/datastore_client.py (hunk 5 above)
Comment (damccorm) [reply playground/infrastructure/datastore_client.py:86]: SGTM

File: examples/java/src/main/java/org/apache/beam/examples/KafkaWordCountAvro.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.transforms.Values;
+import org.apache.beam.sdk.values.KV;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.LongDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+public class KafkaWordCountAvro {
+  static final String TOKENIZER_PATTERN = "[^\\p{L}]+";
```
[the remark this answers is not in the document]
Comment (rshamunov) [reply examples/java/src/main/java/org/apache/beam/examples/KafkaWordCountAvro.java]: Added a comment. We use the same approach for other WordCound examples

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**